### ISSUE TITLE
epoch and batch size agnostic sampler

### ISSUE THREAD

**Issue by MoorayJenkins** [ncullen93/torchsample#1]
Would it be possible to create a sampler that's agnostic to batch size and epoch? Something that seamlessly returns the requested batch size regardless of position. Would help avoiding annoying loops by offloading the problem to the sampler as well as the possibility of injecting noise through random batch size for both the gradient and BN.

`epoch, x, y = sampler.next_batch(bsize=128)`

Ideally it would maintain a sample buffer of configurable size from which it samples and would shuffle in the background so that at the end of an epoch there isn't a noticeable pause.

I have always wanted to write a sampler like that myself but never had the time, but you said you are taking requests.

A more advanced feature could be ensuring that every class is presented an equal number of times within an epoch or even a batch, subject to more transformations to make up for over-representation perhaps.

**Comment by ncullen93** [comment]
Yes like that one tensorflow MNIST example ? That's easy and I just made a design decision not to have that. I'll implement that as an option right now though. The stratified sampling of each class equally would require a new Sampler .. Scikit-learn has a ton of them so I'll either reproduce what they have or maybe just use them directly and add that dependency. Idk yet but ill let you know when I get around to that.

**Comment by MoorayJenkins** [comment]
This is wonderful, thanks.
I was not aware that you reset the iterator internally, I must have overlooked that when I skimmed the code.

One last thing though, it's minor but if you ask for a batch size greater than the number of samples left before a reset it will return an incomplete batch - all it has left over. Maybe have a flag that would guarantee batch size by watching for that special case?


Also there is some odd behavior:

`x = torch.ones(10,1,30,30)
y = torch.from_numpy(np.arange(10))

loader = TensorDataset(x, y, batch_size=3)

xbatch, ybatch = loader.next_batch(5)
print(ybatch.numpy())

xbatch, ybatch = loader.next_batch(7)
print(ybatch.numpy())

xbatch, ybatch = loader.next_batch()
print(ybatch.numpy())`

last fetch causes an error

```
anaconda/lib/python2.7/site-packages/torchsample/dataset_iter.py", line 151, in __next__
    raise StopIteration
StopIteration

```

error goes away with TensorDataset(x, y, batch_size=5) for some reason.


The following

`x = torch.ones(10,1,30,30)
y = torch.from_numpy(np.arange(10))

loader = TensorDataset(x, y, batch_size=6)

xbatch, ybatch = loader.next_batch(2)
print(ybatch.numpy())

xbatch, ybatch = loader.next_batch(3)
print(ybatch.numpy())

xbatch, ybatch = loader.next_batch(2)
print(ybatch.numpy())

xbatch, ybatch = loader.next_batch()
print(ybatch.numpy())`

results in 


[0 1]
[2 3 4]
[0 1]
[2 3 4 5 6 7]


Not clear why it skipped, in some other scenarios it didn't. Just some peculiar behavior.
